### PR TITLE
sleep code refactoring + automatic fallback from hybrid sleep to suspend

### PR DESCRIPTION
**Pull request by poettering**
Started out as a fix for #19550, but then grew into a bunch of other refactoring and fixes.

Fixes: #19550

### CODE REVIEW

File: src/sleep/sleep.c [hunk 1]
```diff
@@ -169,22 +167,42 @@ static int lock_all_homes(void) {
...
+static int execute(
+                const SleepConfig *sleep_config,
+                SleepOperation operation,
+                const char *action) {
+
         char *arguments[] = {
                 NULL,
                 (char*) "pre",
-                arg_verb,
+                /* NB: we use arg_operation instead of operation here, as we wan't to communicate the overall
```
Comment (bluca) [inline src/sleep/sleep.c]: /wan't/want/

**Review by bluca — APPROVED** [review]

**Comment by poettering** [comment]
Force pushed new version, with the one fix @bluca noticed applied. Upgrading green label.

**Comment by bluca** [comment]
Should the new fallback if hybrid sleep fails be documented somewhere?

**Comment by ThomasBrierley** [comment]
This is great :smile: 

I see now that falling back to suspend was already the strategy for `SLEEP_SUSPEND_THEN_HIBERNATE` in `execute_s2h()`.

I was also wondering about when hibernate is used on it's own i.e does there need to be a case added for `SLEEP_HIBERNATE`? that would probably look almost identical to the `SLEEP_HYBRID_SLEEP` case. I realise it is different from the others in that suspend was never part of the intended functionality, but it will still cause the problematic looping as per #19550, would it be a reasonable deviation from it's intended behaviour in the name of safety?

**Comment by poettering** [comment]
> Should the new fallback if hybrid sleep fails be documented somewhere?

Not sure, it's a fallback, i.e. a way of error handling only. We don't document this for s2h failures either, so I wouldn't bother with getting too detailed about this.

**Comment by poettering** [comment]
> I was also wondering about when hibernate is used on it's own i.e does there need to be a case added for `SLEEP_HIBERNATE`? that would probably look almost identical to the `SLEEP_HYBRID_SLEEP` case. I realise it is different from the others in that suspend was never part of the intended functionality, but it will still cause the problematic looping as per #19550, would it be a reasonable deviation from it's intended behaviour in the name of safety?

So in both cases of hybrid sleep and s2h the user basically told us "do both", hence if we fallback from one to the other it's still in line with the general request. However, if people ask us to hibernate and we suspend instead, that's not really inline anymore with what people asked for.

I mean, unless I am mistaken we are not really busy looping again, just trying over and over again with a certain delay in between, which I think is acceptable, since it gives people the chance to do something about.

I think we should polish s2h a bit more so that if people are fine with either suspend or hibernate, then it' the one option you should pick and forget about the rest. Specifically, I'd like to add a mode where we watch battery level in s2h: as long as the battery is above a certain threshold we go to suspend with a timer of 1h or so, but once the battery is below it we go to hibernate instead.

**Comment by ThomasBrierley** [comment]
> [...] if people ask us to hibernate and we suspend instead, that's not really inline anymore with what people asked for.

That's fair.

> unless I am mistaken we are not really busy looping again, just trying over and over again with a certain delay in between, which I think is acceptable, since it gives people the chance to do something about.

Yup, there is a small delay. Perhaps this is not as common a scenario as I think. My problem was that it still took a while to catch it due to the temperamental nature of hibernation with under-provisioned swap... so it will start failing when you aren't necessarily looking for it. Laptops are quiet enough these days that failure is not immediately obvious.

Anyway... I'm picking holes. I think your fix covers the vast majority of use cases, thanks :smile: I started patching the man page but if "hibernate" is not going to be included, then yeah it feels unnecessary.

> Specifically, I'd like to add a mode where we watch battery level in s2h: as long as the battery is above a certain threshold we go to suspend with a timer of 1h or so, but once the battery is below it we go to hibernate instead.

That would be a nice improvement, in fact s2h failing on low battery is the only reason I continue to use "hybrid-sleep".

**Pull request merged**